RecoBTag/ONNXRuntime/plugins/ParticleTransformerAK4SonicJetTagsProducer.cc

```diff
@@ -0,0 +1,306 @@
+    : TritonEDProducer<>(iConfig),
+      src_(consumes<TagInfoCollection>(iConfig.getParameter<edm::InputTag>("src"))),
+      flav_names_(iConfig.getParameter<std::vector<std::string>>("flav_names")),
+      input_names_(iConfig.getParameter<std::vector<std::string>>("input_names")),
+      output_names_(iConfig.getParameter<std::vector<std::string>>("output_names")) {
+  // get output names from flav_names
+  for (const auto& flav_name : flav_names_) {
+    produces<JetTagCollection>(flav_name);
+  }
+}
+
+ParticleTransformerAK4SonicJetTagsProducer::~ParticleTransformerAK4SonicJetTagsProducer() {}
+
+void ParticleTransformerAK4SonicJetTagsProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // pfParticleTransformerAK4JetTags
+  edm::ParameterSetDescription desc;
+  TritonClient::fillPSetDescription(desc);
+  desc.add<edm::InputTag>("src", edm::InputTag("pfParticleTransformerAK4TagInfos"));
+  desc.add<std::vector<std::string>>("input_names", {"input_1", "input_2", "input_3", "input_4", "input_5", "input_6"});
+  desc.add<std::vector<std::string>>("output_names", {"softmax"});
+  desc.add<std::vector<std::string>>(
+      "flav_names", std::vector<std::string>{"probb", "probbb", "problepb", "probc", "probuds", "probg"});
+
+  descriptions.add("pfParticleTransformerAK4SonicJetTags", desc);
+}
+
+void ParticleTransformerAK4SonicJetTagsProducer::acquire(edm::Event const &iEvent, edm::EventSetup const &iSetup, Input &iInput) {
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+  client_->setBatchSize(tag_infos->size());
+  skippedInference_ = false;
+  if (tag_infos->empty()) return;
+  unsigned int max_n_cpf = 0; 
+  unsigned int max_n_npf = 0; 
+  unsigned int max_n_vtx = 0; 
+
+  // Find the max n_cpf, n_npf and n_vtx among all the jets in an event. 
+  for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+    max_n_cpf = std::max(max_n_cpf,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().c_pf_features.size()));
+    max_n_npf = std::max(max_n_npf,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().n_pf_features.size()));
+    max_n_vtx = std::max(max_n_vtx,
+                         static_cast<unsigned int>(((*tag_infos)[jet_n]).features().sv_features.size()));
+  }
+  
+  // If an event has no jet, or all jets has zero n_cpf, n_npf and n_vtx, the inference is skipped.
+  if (max_n_cpf == 0 && max_n_npf == 0 && max_n_vtx == 0) {
+    client_->setBatchSize(0);
+    skippedInference_ = true;
+    return;
+  }
+    
+  // all the jets in the same event will fill up the same amount of n_cpf, n_npf, n_vtx and send to server
+  max_n_cpf = std::clamp(max_n_cpf, (unsigned int)0, (unsigned int)25);
+  max_n_npf = std::clamp(max_n_npf, (unsigned int)0, (unsigned int)25);
+  max_n_vtx = std::clamp(max_n_vtx, (unsigned int)0, (unsigned int)5);
```

Comment:
these magic numbers should be named constants, both here and in the direct inference producer. (also, you can put the use of `clamp()` into the direct inference producer)

Reply:
made code change, please check